### PR TITLE
Add description field into API and web interface

### PR DESCRIPTION
**Pull request by kapsh**

### CODE REVIEW

**Comment by kapsh** [comment]
1. Optional description can be set in both web interface and api request.
2. Will be shown when available in web.
3. Will be empty in api response when wasn't set (last thing people want is KeyError really).

Tested this locally.

Now to sort out how to implement this
> <ikanobori> How about if the URL returned by the API would also include the filename behind a #?

File: pinnwand/http.py [hunk 1]
```diff
@@ -86,7 +89,8 @@ def check_xsrf_cookie(self) -> None:
 
 
 class ShowPaste(Base):
-    async def get(self, paste_id: str) -> None:
+    async def get(self, path: str) -> None:
+        paste_id, _, description = path.partition("#")  # ignore anchor part
```
Comment (kapsh) [inline pinnwand/http.py]: What would be correct way to process url here? Scanned tornado.routing docs, didn't found any.

File: pinnwand/http.py [hunk 2]
```diff
@@ -198,14 +206,23 @@ class APINew(Base):
...
+        paste = database.Paste(
+            raw, lexer, utility.expiries[expiry], description
+        )
 
         with database.session() as session:
             session.add(paste)
             session.commit()
 
+            req_url = self.request.full_url()
+            location = f"{paste.paste_id}#{url_escape(paste.description)}"
```
Comment (kapsh) [inline pinnwand/http.py]: Also asking for better way.

**Comment by kapsh** [comment]
Please review and maybe give me some hints on questions above.

**Comment by supakeen** [comment]
I'll review this on Sunday! The idea is good but perhaps the name 'description' is a bit more wide than what we actually want, a filename.

**Comment by kapsh** [comment]
I don't like "description" either for being longer than other request parameters. What do you like more, "name" or "title"?

File: pinnwand/http.py (hunk 1 above)
Comment (kapsh) [reply pinnwand/http.py]: Uh, did some regex magic in urlspec.

File: pinnwand/http.py (hunk 2 above)
Comment (supakeen) [reply pinnwand/http.py]: This is the correct way to do this. You could roll this into line 222 and 223 if you want so you don't have extra assignments but this is perfectly readable.

File: pinnwand/http.py [hunk 3]
```diff
@@ -264,8 +280,8 @@ def make_application() -> tornado.web.Application:
         [
             (r"/", CreatePaste),
             (r"/\+(.*)", CreatePaste),
-            (r"/show/(.*)", ShowPaste),
-            (r"/raw/(.*)", RawPaste),
+            (r"/show/(.*)(?:#.+)?", ShowPaste),
+            (r"/raw/(.*)(?:#.+)?", RawPaste),
             (r"/remove/(.*)", RemovePaste),
```
Comment (supakeen) [inline pinnwand/http.py]: If you add these they should be added to APIShow/etc as well.

File: pinnwand/template/new.html [hunk 4]
```diff
@@ -9,6 +9,10 @@
 <form action="/" method="post" class="create">
     <textarea name="code" spellcheck="false" placeholder="Enter code to paste."></textarea>
 
+    <div class="extra">
+    Description: <input type="text" name="description" size="50"/>
+    </div>
+
```
Comment (supakeen) [inline pinnwand/template/new.html]: Let's leave description out for any web pastes.

**Comment by kapsh** [comment]
>  you could add a `.url` property to the Paste class so you don't keep having to create the URLs on the fly where they are used!

By url you mean full url or `f'/{id}#{description}'` part? Storing former doesn't look like good idea because domain running pinnwand instance could change eventually, breaking stored links.

File: pinnwand/http.py [hunk 5]
```diff
@@ -183,6 +189,7 @@ class APINew(Base):
         lexer = self.get_body_argument("lexer")
         raw = self.get_body_argument("code")
         expiry = self.get_body_argument("expiry")
+        description = self.get_body_argument("description", "")
```
[the remark this answers is not in the document]
Comment (kapsh) [reply pinnwand/http.py]: Noned it. This made constructing location slightly longer, please check line 217.

**Comment by kapsh** [comment]
Sorry, I know the small changes rule, but I've messed my local commits while rebasing and decided to squash them into one.

**Comment by supakeen** [comment]
That's ok, I'll review this again after work.

**Comment by kapsh** [comment]
Merge conflict resolved.

**Comment by supakeen** [comment]
The last bit that would make this mergeable for me is if you rename `description` to `filename`. That's more what it actually is.

If you can then preferably add a testcase that sees if it gets returned from the API it'll be an instant merge :)

**Comment by kapsh** [comment]
Testcase was really useful indeed, it has catched one place where I forgot to rename parameter. 

Also I've tried to add some description on API usage to readme file. By the way, while documenting I found a couple of possible inconveniences (spelling things always helps).
1. Perhaps APIRemove is better to use `/json/remove/removal_id` request without body. Looks more REST-ish.
2. `/json/new` really needs only `code` field, others can use application default if weren't supplied. 

I'll leave these ideas on your consideration.

**Comment by supakeen** [comment]
This looks perfectly fine now. As for API changes: no, it's currently already being used so changing things like the removal route would require a new endpoint instead of clobbering the old (there's more examples and comments in the code regarding backwards compatibility).

If you want you can add your name to the AUTHORS.rst file and I can merge this, if you don't want to then I will merge after you let me know!

**Comment by kapsh** [comment]
Thank you, I'm fine with merging as it is.

**Comment by supakeen** [comment]
Merged, I don't do continuous deployment so deployment will follow somewhere in the next few days together with another change I have planned :)

**Comment by kapsh** [comment]
@supakeen it's okay, I would only ask to ping me back after you do it. I will add this functionality in my pet pastebin client I'm working on.

**Pull request merged**